src/repr/src/row.proto

```diff
@@ -0,0 +1,135 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+syntax = "proto3";
+
+package gen;
+
+message ProtoRow {
+    repeated ProtoDatum datums = 1;
+}
+
+message ProtoDatum {
+    // NB: Proto oneof ids `1..=15` get encoded in 1 byte and so we should
+    // reserve them for the datum types we expect to be most popular.
+    oneof datum_type {
+        ProtoDatumOther other = 1;
```

Comment:
It'd be a little more efficient to use empty variants, no?

```diff
diff --git a/src/repr/src/row.proto b/src/repr/src/row.proto
index 76f0a86b7..3366981da 100644
--- a/src/repr/src/row.proto
+++ b/src/repr/src/row.proto
@@ -9,6 +9,8 @@
 
 syntax = "proto3";
 
+import "google/protobuf/empty.proto";
+
 package gen;
 
 message ProtoRow {
@@ -19,14 +21,17 @@ message ProtoDatum {
     // NB: Proto oneof ids `1..=15` get encoded in 1 byte and so we should
     // reserve them for the datum types we expect to be most popular.
     oneof datum_type {
-        ProtoDatumOther other = 1;
-        int32 int16 = 2;
-        int32 int32 = 3;
-        int64 int64 = 4;
-        float float32 = 5;
-        double float64 = 6;
-        bytes bytes = 7;
-        string string = 8;
+        google.protobuf.Empty null = 1;
+        google.protobuf.Empty json_null = 2;
+        google.protobuf.Empty dummy = 3;
+        bool bool = 4;
+        int32 int16 = 5;
+        int32 int32 = 6;
+        int64 int64 = 7;
+        float float32 = 8;
+        double float64 = 9;
+        bytes bytes = 10;
+        string string = 11;
 
         // These get encoded with 2 bytes for the oneof id. It's a pretty easy
         // and low-debt migration to "bless" one of these into having a 1-byte
@@ -48,22 +53,6 @@ message ProtoDatum {
     }
 }
 
-enum ProtoDatumOther {
-    // It's generally good practice to make id 0 (the default if the field is
-    // unset) in proto enums be an Unknown sentinel. This allows for
-    // distinguishing between unset and any of the enum variants.
-    //
-    // This enum is initially used only in a oneof, which means we can
-    // distinguish unset without this sentinel. But stick one in here anyway,
-    // in case this enum gets used somewhere else in the future.
-    Unknown = 0;
-    Null = 1;
-    False = 2;
-    True = 3;
-    JsonNull = 4;
-    Dummy = 5;
-}
-
 message ProtoDate {
     // Year
     int32 year = 1;
diff --git a/src/repr/src/row/encoding.rs b/src/repr/src/row/encoding.rs
index f8504c084..ece8d7f47 100644
--- a/src/repr/src/row/encoding.rs
+++ b/src/repr/src/row/encoding.rs
@@ -16,13 +16,14 @@ use dec::Decimal;
 use ore::cast::CastFrom;
 use persist_types::Codec;
 use protobuf::{Message, MessageField};
+use protobuf::well_known_types::Empty;
 use uuid::Uuid;
 
 use crate::adt::array::ArrayDimension;
 use crate::adt::interval::Interval;
 use crate::gen::row::proto_datum::Datum_type;
 use crate::gen::row::{
-    ProtoArray, ProtoArrayDimension, ProtoDate, ProtoDatum, ProtoDatumOther, ProtoDict,
+    ProtoArray, ProtoArrayDimension, ProtoDate, ProtoDatum, ProtoDict,
     ProtoInterval, ProtoNumeric, ProtoRow, ProtoTime, ProtoTimestamp,
 };
 use crate::{Datum, Row};
@@ -64,8 +65,8 @@ impl Codec for Row {
 impl<'a> From<Datum<'a>> for ProtoDatum {
     fn from(x: Datum<'a>) -> Self {
         let datum_type = match x {
-            Datum::False => Datum_type::other(ProtoDatumOther::False.into()),
-            Datum::True => Datum_type::other(ProtoDatumOther::True.into()),
+            Datum::False => Datum_type::bool(false),
+            Datum::True => Datum_type::bool(true),
             Datum::Int16(x) => Datum_type::int16(x.into()),
             Datum::Int32(x) => Datum_type::int32(x),
             Datum::Int64(x) => Datum_type::int64(x),
@@ -160,10 +161,10 @@ impl<'a> From<Datum<'a>> for ProtoDatum {
                     cached_size: Default::default(),
                 })
             }
-            Datum::JsonNull => Datum_type::other(ProtoDatumOther::JsonNull.into()),
+            Datum::JsonNull => Datum_type::json_null(Empty::new()),
             Datum::Uuid(x) => Datum_type::uuid(x.as_bytes().to_vec()),
-            Datum::Dummy => Datum_type::other(ProtoDatumOther::Dummy.into()),
-            Datum::Null => Datum_type::other(ProtoDatumOther::Null.into()),
+            Datum::Dummy => Datum_type::dummy(Empty::new()),
+            Datum::Null => Datum_type::null(Empty::new()),
         };
         ProtoDatum {
             datum_type: Some(datum_type),
@@ -187,15 +188,11 @@ impl From<&Row> for ProtoRow {
 impl Row {
     fn try_push_proto(&mut self, x: &ProtoDatum) -> Result<(), String> {
         match &x.datum_type {
-            Some(Datum_type::other(o)) => match o.enum_value() {
-                Ok(ProtoDatumOther::Unknown) => return Err("unknown datum type".into()),
-                Ok(ProtoDatumOther::Null) => self.push(Datum::Null),
-                Ok(ProtoDatumOther::False) => self.push(Datum::False),
-                Ok(ProtoDatumOther::True) => self.push(Datum::True),
-                Ok(ProtoDatumOther::JsonNull) => self.push(Datum::JsonNull),
-                Ok(ProtoDatumOther::Dummy) => self.push(Datum::Dummy),
-                Err(id) => return Err(format!("unknown datum type: {}", id)),
-            },
+            Some(Datum_type::null(_)) => self.push(Datum::Null),
+            Some(Datum_type::json_null(_)) => self.push(Datum::JsonNull),
+            Some(Datum_type::dummy(_)) => self.push(Datum::Dummy),
+            Some(Datum_type::bool(false)) => self.push(Datum::False),
+            Some(Datum_type::bool(true)) => self.push(Datum::True),
             Some(Datum_type::int16(x)) => {
                 let x = i16::try_from(*x)
                     .map_err(|_| format!("int16 field stored with out of range value: {}", *x))?;

```

Reply:
We can definitely move true/false back in. I'll do that.

The reason I didn't do the others is because I know if it was worth one of our precious 1-byte ids for things like Dummy and JsonNull. Now that you bring up the Empty proto idea, I'm not sure it saves us anything. I'll double check, but I'm pretty sure this would be encoded as a 1 byte varint for the set field id+type followed by one that's always 0 for the length of the Empty message.

Reply:
Okay wrote a big long comment that described each individual byte in the actual encoding output of various options here, but github lost it. Summary is that I was correct that we have to encode the length of the Empty message, it's not special cased, so it doesn't get us anything. Great idea tho, I'm glad we checked it out

Along the way, I also realized that we end up using the same total length for things like Null, True, False that are representable as a proto enum variant (i.e. no payload), so I've reverted the bit since it needlessly uses our precious datum_type oneof ids.

Reply:
Gotcha sounds good!